communication-refactored/src/interface.rs

```diff
@@ -0,0 +1,559 @@
+pub use types::*;
+
+#[derive(Clone)]
+/// Interface for the stronghold-communication library to create a swarm, handle events and perform operations.
+///
+/// All Swarm interaction takes place in a separate task.
+/// [`ShCommunication`] is essentially a wrapper for the Sender side of a mpsc channel, which is used to initiate
+/// operations on the swarm.
+///
+/// Refer to [`ShCommunicationBuilder`] for more information on the default configuration.
+pub struct ShCommunication<Rq, Rs, TRq = Rq>
+where
+    // Request message type
+    Rq: RqRsMessage + Borrow<TRq>,
+    // Response message type
+    Rs: RqRsMessage,
+    // Optional, tailored request-type that is used in the firewall to get approval.
+    // This has the purpose of trimming the actual request down to the firewall-relevant information and e.g. avoid
+    // exposing the request's actual content.
+    TRq: Clone + Send + 'static,
+{
+    // Id of the local peer.
+    local_peer_id: PeerId,
+    // Channel for sending [`SwarmOperation`] to the [`SwarmTask`] .
+    // The [`SwarmOperation`]s trigger according operations on the Swarm.
+    // The result of an operation is received via the oneshot Receiver that is included in each type.
+    command_tx: mpsc::Sender<SwarmOperation<Rq, Rs, TRq>>,
+}
+
+impl<Rq, Rs, TRq> ShCommunication<Rq, Rs, TRq>
+where
+    Rq: RqRsMessage + Borrow<TRq>,
+    Rs: RqRsMessage,
+    TRq: Clone + Send + 'static,
+{
+    /// Create a new [`ShCommunication`] instance with the default configuration.
+    /// Refer to [`ShCommunicationBuilder::new`] and [`ShCommunicationBuilder::build`] for detailed information.
+    #[cfg(feature = "tcp-transport")]
+    pub async fn new(
+        firewall_channel: mpsc::Sender<FirewallRequest<TRq>>,
+        requests_channel: mpsc::Sender<ReceiveRequest<Rq, Rs>>,
+        events_channel: Option<mpsc::Sender<NetworkEvent>>,
+    ) -> Result<Self, io::Error> {
+        ShCommunicationBuilder::new(firewall_channel, requests_channel, events_channel)
+            .build()
+            .await
+    }
+
+    /// Get the [`PeerId`] of the local peer.
+    pub fn get_peer_id(&self) -> PeerId {
+        self.local_peer_id
+    }
+
+    /// Send a new request to a remote peer.
+    ///
+    /// This will attempt to establish a connection to the remote via one of the known addresses, if there is no active
+    /// connection.
+    pub async fn send_request(&mut self, peer: PeerId, request: Rq) -> Result<Rs, OutboundFailure> {
+        let (tx_yield, rx_yield) = oneshot::channel();
+        let command = SwarmOperation::SendRequest {
+            peer,
+            request,
+            tx_yield,
+        };
+        self.send_command(command).await;
+        rx_yield.await.unwrap()
+    }
+
+    /// Start listening on the network.
+    /// If no address is given, the listening address will be OS-assigned.
+    pub async fn start_listening(&mut self, address: Option<Multiaddr>) -> Result<Multiaddr, ListenErr> {
```

Comment:
When e.g. listening on `/ip4/0.0.0.0/tcp/0` multiple listen addresses might be returned, no?

Reply:
Yes that's true. The problem is that I don't see a way of determining when the last "new-listen-address" is reported, hence when I have "collected" all addresses associated with a new listener. I am currently only returning the address from the first reported `SwarmEvent::NewListenAddr` of the associated Listener.
Do you see a way of solving this?

But either way, I will correct my comment, since it currently implies that there is only one listening address, which is not correct in cases like the one you mentioned.

Reply:
Given that a listener might emit a `NewAddress` at any point in time, the only way I see to properly report this is a channel instead of a oneshot, that lives as long as the listener, being used for each `NewListenAddr`.

That said, I am not sure it is worth the complexity in your use-case.